### ISSUE TITLE
[HOLD for payment 2023-03-09] [$1000] Wrong validation error message on choose date of birth.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15228]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open Settings -> Profile -> Personal Details.
2. Click on Date of birth.
3. Choose date 17/2/2018.( Testing this scenario on 02/16)
4. Press on save button.
5. Error message will be display "Date should be before 2018-02-16."
6. Now select 16/2/2018, it will works.

## Expected Result:
Either message should be "Date should be on or before 2018-02-16.". Or don't allow to select 16/2/2018

## Actual Result:
Incorrect error message allowing to select 16/2/2018 but error message says date should be before 16/2/2018

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.72-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/219496871-45289008-2ee5-41dc-af2a-618074785908.mov

https://user-images.githubusercontent.com/43996225/219496894-b02110a0-cdc7-4320-aec8-d4d741277e1e.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @jatinsonijs 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1676539597780339

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01d2cc56f9c3202414</li>
        <li>Upwork Job ID: 1626535149478653952</li>
        <li>Last Price Increase: 2023-02-17</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @dylanexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by dylanexpensify** [comment]
Cannot reproduce on staging, @kavimuru can you share what platforms this occurs on and if still occurring? 🙇

**Comment by dylanexpensify** [comment]
Was able to reproduce, but we need to update steps to say set to _today's date_ not Feb 17 cc @kavimuru

**Comment by dylanexpensify** [comment]
[Thread](https://expensify.slack.com/archives/C049HHMV9SM/p1676539597780339)

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01d2cc56f9c3202414

**Comment by MelvinBot** [comment]
Current assignee @dylanexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @danieldoglas (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by jatinsonijs** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Validation message is incorrect or we can say incomplete

### What is the root cause of that problem?
We are allowing ages 5 to 150 so we are doing some validation to check age is between these numbers or not. but message is incorrect in translation: https://github.com/Expensify/App/blob/90b9efa37819b823c133f35ec6b7d141ebdd3374/src/languages/en.js#L617

### What changes do you think we should make in order to solve the problem?
We have to just change message in translation to indicate before or equal to `date `
and same for after date - after or equal to `date`

**Comment by situchan** [comment]
Should we put this on HOLD for [New Date Picker Design](https://github.com/Expensify/App/issues/14322)?

**Comment by allroundexperts** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The validation error message says that we can not select a particular date during DOB selection but in actual, we can select that date without any error.

### What is the root cause of that problem?
The root cause of this problem is that current date that we use for comparison includes hours of the day as well. The selected date however, does not include the hours of the day. eg 18-01-01 translates to `01 Jan 2018 00:00:00`. This causes the check to pass as this date is before the current date.

### What changes do you think we should make in order to solve the problem?

We can solve the problem by making sure that we're comparing against the current date with start of the day selected. The following needs to be changed:
https://github.com/Expensify/App/blob/main/src/libs/ValidationUtils.js#L218-L219
to
```
    const recentDate = moment().startOf('day').subtract(minimumAge, 'years');
    const longAgoDate = moment().startOf('day').subtract(maximumAge, 'years');
```
We also need to change the following:
https://github.com/Expensify/App/blob/main/src/libs/ValidationUtils.js#L227
to
```
    if (testDate.isSameOrAfter(recentDate)) {
```

### What alternative solutions did you explore? (Optional)
- Changing the validation message. This however does not consider the root cause.

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by dylanexpensify** [comment]
@rushatgabhane can we get a review of @allroundexperts proposal please?

**Comment by danieldoglas** [comment]
Makes sense to me. Assigning @allroundexperts

**Comment by MelvinBot** [comment]
📣 @allroundexperts You have been assigned to this job by @danieldoglas!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by allroundexperts** [comment]
@rushatgabhane @danieldoglas PR created https://github.com/Expensify/App/pull/15316

**Comment by MelvinBot** [comment]
@danieldoglas, @rushatgabhane, @allroundexperts, @dylanexpensify Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by danieldoglas** [comment]
Wait up melvin, this is in staging.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.77-4 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15316


If no regressions arise, payment will be issued on 2023-03-09. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@rushatgabhane / @danieldoglas] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@rushatgabhane / @danieldoglas] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@rushatgabhane / @danieldoglas] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@allroundexperts] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@dylanexpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by dylanexpensify** [comment]
Reviewing today!

**Comment by danieldoglas** [comment]
@rushatgabhane can you please fill the checklist? @allroundexperts also please propose regression tests on this.

**Comment by allroundexperts** [comment]
**Regression Test Proposal**

1. Go to https://staging.new.expensify.com/ and open Settings -> Profile -> Personal Details.
2. Click on Date of birth.
3. Enter a date that is exactly 5 years before the current date.
4. Press save button.
5. Verify that the app shows an error message.
6. Enter another date which is 5 years and 1 day before current date.
7. Verify that no error appears.
Do we agree 👍 or 👎

**Comment by MelvinBot** [comment]
Triggered auto assignment to @dylanexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by dylanexpensify** [comment]
lol

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mallenexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by dylanexpensify** [comment]
@mallenexpensify I'm heading OOO! Would you mind finishing the last bits for me?? 🙇

**Comment by mallenexpensify** [comment]
@jatinsonijs @rushatgabhane @allroundexperts can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01d2cc56f9c3202414

This bug is specific to the date being exactly 5 years from the current date, right?

**Comment by allroundexperts** [comment]
> @jatinsonijs @rushatgabhane @allroundexperts can you please accept the job and reply here once you have? https://www.upwork.com/jobs/~01d2cc56f9c3202414
> 
> This bug is specific to the date being exactly 5 years from the current date, right?

Currently we have it set up for 5 years.

**Comment by allroundexperts** [comment]
I've applied @mallenexpensify.

**Comment by mallenexpensify** [comment]
> Currently we have it set up for 5 years.

Thanks @allroundexperts .   @isabelastisser , when you're back from ECX, can you take a look at this? I'm unsure it should be added to TestRail because it's an extreme edge case that anyone (outside of @jatinsonijs , ha) would add a birthday of exactly 5 years.

**Comment by jatinsonijs** [comment]
@mallenexpensify applied

**Comment by mallenexpensify** [comment]
@isabelastisser I swear I updated it right after I posted it! (honestly, I probably have done it more than once this week, sorry)

**Comment by mallenexpensify** [comment]
Everyone is paid!!  Leaving open and bumping to weekly til @isagoico  is back and can review.

**Comment by allroundexperts** [comment]
@mallenexpensify Thanks for the payment. The PR for this was merged within 3 days. Am I eligible for the bonus here?

**Comment by allroundexperts** [comment]
> @mallenexpensify Thanks for the payment. The PR for this was merged within 3 days. Am I eligible for the bonus here?

I take my comment back. This was merged ~12 hours late.

**Comment by isagoico** [comment]
> > Currently we have it set up for 5 years.
> 
> Thanks @allroundexperts . @isagoico , when you're back from ECX, can you take a look at this? I'm unsure it should be added to TestRail because it's an extreme edge case that anyone (outside of @jatinsonijs , ha) would add a birthday of exactly 5 years.

Lol agree here, I don't think this would be a valuable addition since there's a extremely low chance of this happening to any user. So I would say :donothing: in TestRail for this case.

**Comment by mallenexpensify** [comment]
Thanks for confirming @isagoico , closing
